source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+            right_cyl_r_item.index_count = right_cyl_r_item.geometry->draw_args["cylinder"_small].index_count;
+            right_cyl_r_item.start_index_location = right_cyl_r_item.geometry->draw_args["cylinder"_small].start_index_location;
+            right_cyl_r_item.base_vertex_location = right_cyl_r_item.geometry->draw_args["cylinder"_small].base_vertex_location;
+            right_cyl_r_item.num_frames_dirty = renderer::num_frames_in_flight();
+
+            left_sphere_r_item.world = left_sphere_world;
+            left_sphere_r_item.constant_buffer_index = obj_cb_index++;
+            left_sphere_r_item.geometry = g_regina_ctx.meshes["scene_geometry"_med].get();
+            left_sphere_r_item.topology = renderer::PrimitiveTopology::TRIANGLELIST;
+            left_sphere_r_item.index_count = left_sphere_r_item.geometry->draw_args["sphere"_small].index_count;
+            left_sphere_r_item.start_index_location = left_sphere_r_item.geometry->draw_args["sphere"_small].start_index_location;
+            left_sphere_r_item.base_vertex_location = left_sphere_r_item.geometry->draw_args["sphere"_small].base_vertex_location;
+            left_sphere_r_item.num_frames_dirty = renderer::num_frames_in_flight();
+
+            right_sphere_r_item.world = right_sphere_world;
+            right_sphere_r_item.constant_buffer_index = obj_cb_index++;
+            right_sphere_r_item.geometry = g_regina_ctx.meshes["scene_geometry"_med].get();
+            right_sphere_r_item.topology = renderer::PrimitiveTopology::TRIANGLELIST;
+            right_sphere_r_item.index_count = right_sphere_r_item.geometry->draw_args["sphere"_small].index_count;
+            right_sphere_r_item.start_index_location = right_sphere_r_item.geometry->draw_args["sphere"_small].start_index_location;
+            right_sphere_r_item.base_vertex_location = right_sphere_r_item.geometry->draw_args["sphere"_small].base_vertex_location;
+            right_sphere_r_item.num_frames_dirty = renderer::num_frames_in_flight();
+
+            g_regina_ctx.scene->add_render_item(rsl::move(left_cyl_r_item));
+            g_regina_ctx.scene->add_render_item(rsl::move(right_cyl_r_item));
+            g_regina_ctx.scene->add_render_item(rsl::move(left_sphere_r_item));
+            g_regina_ctx.scene->add_render_item(rsl::move(right_sphere_r_item));
+        }
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+    bool build_frame_resources()
+    {
+        for (int i = 0; i < renderer::num_frames_in_flight(); ++i)
+        {
+            renderer::ResourceSlot frame = renderer::create_frame_resource();
+
+            g_regina_ctx.frame_resource_data.push_back({rsl::move(frame) , renderer::ResourceSlot::make_invalid(), renderer::ResourceSlot::make_invalid()});
+        }
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+    bool build_constant_buffers()
```

Comment:
I find this function very confusing for what it's supposed to do.
you create 2 buffers, one containing an array of per object constant buffers and one containing a per pass constant buffer. that's the first for loop, I get that.

The second and third loop though are much unclear how it works. I know it works in the end, because I see something on screen, but I had to deep dive into how it all works. You're creating view per render item, per frame, so far is clear, but I couldn't figure out how you made sure that the cb of 1 object is actually properly stored and you can get the view of it.
I suggest to write comments explaining how it works or refactor a bit so it's clearer how it all works together

Reply:
After thinking about it a bit more the first loop can actually be stored within the call to create a new constant buffer. I don't know if you can create an uploader and repurpose it for different item but at least I can store the "CommittedResource" (which will hold a reference to the uploader) within the resource pool when a constant buffer is allocated. This is internal knowledge that should not be exposed. I will add another issue for this ( added some comments for now )